### ISSUE TITLE
[RFC] Moving the Octicons source of truth into GitHub

### ISSUE THREAD

**Issue by colebemis** [primer/octicons#372]
## TL;DR

I’m proposing that we move the Octicons source of truth from Figma into GitHub because it would make icon changes easier to track, preview, and publish. 

## Problem

We currently have a [master Figma file](https://www.figma.com/file/FP7lqd1V00LUaT5zvdklkkZr/Octicons?node-id=24251%3A2) that is the source of truth for Octicons. When we push a commit to the master branch of the [primer/octicons](https://github.com/primer/octicons) repo, a GitHub action exports SVG data from the master Figma file and uses that data to build various packages. Here’s a diagram of the current setup:

![Source of truth](https://user-images.githubusercontent.com/4608155/74988416-a265e800-53f2-11ea-8a9e-cba4b1734bf0.png)

> You can learn more about how we currently use Figma in [this blog post](https://github.blog/2018-04-12-driving-changes-from-designs/) by @jonrohan. 

This setup is nice because contributors don’t need experience with git or a local development setup to contribute. However, there are a few drawbacks:

### 1. Icons are not version controlled 
It’s difficult to see how an icon has changed over time or recover an icon if it accidentally gets deleted. Figma has a “Version History” feature but it’s capabilities are pretty limited compared to git.

### 2.  Mistakes can easily go unnoticed
Changes to the master Figma file don't go through the same review process that code does. To illustrate the issue, let’s imagine that a maintainer accidentally deletes an icon while updating the master Figma file. The next time someone pushes a commit to the master branch of [primer/octicons](https://github.com/primer/octicons) that icon would be removed from all the packages without anyone noticing. 

### 3. Contributors don’t get recognized
Here’s how contributions currently work: A contributor begins the process by opening an issue with a link to a Figma file that contains the icon they’d like to contribute. If approved, a maintainer copies the SVG from the given Figma file into the master Figma file and closes the issue. Since no code was merged into the repo, the contributor is not recognized in the “contributors” list on the GitHub repo. 

### 4. It’s hard to preview changes
We currently rely on before/after screenshots and Figma files to preview icon changes. However, it’s hard to understand the impact of a change without seeing the icon used in context.

### 5. Releasing new icons requires many manual steps for maintainers
Here’s how the release process currently works: After an icon has been approved, a maintainer adds the icon to the master Figma file. Then the maintainer creates a release branch, updates the version number, and merges the branch if all the checks pass. This process is not too hard but it leaves a lot of room for human error and, as I mentioned before, mistakes can easily go unnoticed.

## Proposed solution

I’m proposing that we move the source of truth from Figma into GitHub. The icon SVGs would live in the GitHub repo and be used to build various packages, a Figma plugin, and the Octicons website. To contribute, people would open pull requests adding or updating SVG files. Here’s a diagram of the architecture I’m purposing:

![Source of truth](https://user-images.githubusercontent.com/4608155/74988454-bad60280-53f2-11ea-8111-015a9464fcf1.png)

With this setup, icon changes would be easier to track, preview, and publish:

### 📝 Easier to track
- Icons would be saved as version-controlled SVG files so we’d have a record of every change that’s been made to each icon.
- We’d get an accurate representation of contributions because we’d merge pull requests instead of closing them.

### 👀 Easier to preview
- Contributors and maintainers could easily preview changes using GitHub’s built-in image diff viewer or with deploy previews.

### 🚢 Easier to publish
- The release process would be similar to our other open source libraries and could be done entirely without leaving GitHub.

## Questions

* As part of my proposal, we'd replace the Octicons Figma component library with a Figma plugin (similar to the [Feather Figma plugin](https://www.figma.com/community/plugin/744047966581015514/Feather-Icons)). Are there any drawbacks to using a Figma plugin instead of a Figma component library?

***

I’d love to hear what people think of this idea 😺

cc @primer/design-systems

**Comment by superbryntendo** [comment]
💯💯💯

 I can not voice my support for this strongly enough. The figma file is useful for placement in mocks, but really falls short when it comes to contribution.

**Comment by yaili** [comment]
This is a blog post ready to be published! 🤗🚢

**Comment by ashygee** [comment]
I agree with @superbryntendo and think this is a great proposal! If we go through this route we will need to update our contribution guidelines to include the necessary requirements for the SVGs (outlined strokes, all shapes combined (union selection), etc.) but that won't be hard at all. I really love how this will push forward contribution credit too. We've had so many great contributions but really only one person is publicly seen merging it all together. 

For the plugin vs component library.. ideally I'd like to have both. It would be great if the plugin could be similar to the [proposal](https://pxlc.at/post/ck6mtmz2qig1i07703yzcjz28) I made where icons are components that are replaced but they all still live together in a unified library. I prefer this because there are lots of ways to access a component library, whether via the Assets tab or the instance dropdown or having the library file open. Having them in the library will allow for them to be more accessible across Figma IMO.

**Comment by joelhawksley** [comment]
@colebemis thanks for writing this up! I have no objections 👍 

I'd love to hear what the Figma folks have to say regarding this feedback. Perhaps they might have a workflow that is more conducive to the goals you outlined here?

**Comment by colebemis** [comment]
> For the plugin vs component library.. ideally I'd like to have both. It would be great if the plugin could be similar to the proposal I made where icons are components that are replaced but they all still live together in a unified library.

Is it possible to update a Figma component library automatically from code? As far as I know, it’s not. I’d like to avoid have two sources of truth (i.e. Figma library and GitHub repo) if possible.

**Comment by superbryntendo** [comment]
It seems to me that it might be easier to bypass figma's component library altogether and use a simple microsite that allows for copying glyphs as svgs and then just pasting them into figma as needed. This has the added benefit of making them easily accessible for non-designers as well. Otherwise, could use a plugin like @brianlovin's data populator to pull svg's from an api and generate frames for each that can then be used via figma's component library.

**Comment by colebemis** [comment]
@superbryntendo Yep, we'll build a microsite for copying SVGs no matter what. It'll be similar to https://primer.style/octicons-v2. 

What if we start by building an Octicons Figma plugin that just inserts the SVG directly into the document (like the [Feather plugin](https://www.figma.com/community/plugin/744047966581015514/Feather-Icons) does)? If we find that we're missing any functionality that'd we get from Figma components, we can build another Figma plugin to generate the component library page. How does that sound @ashygee?

**Comment by ashygee** [comment]
@colebemis if everyone feels that's best I'm all for it. I just want to ensure that swapping the icons will be easy. Copying and pasting svgs doesn't seem as efficient to me but I'm happy to test.

**Comment by colebemis** [comment]
> I just want to ensure that swapping the icons will be easy.

I think we can handle swapping with the Figma plugin 👍
